src/math/ring_signature.cpp

```diff
@@ -0,0 +1,363 @@
+            const auto& s = out.s[i][j];
+
+            // Calculate e and R until the end of this ring.
+            const auto e_i_j = borromean_hash(M, R_i_j, i, j);
+            R_i_j = calculate_R(s, e_i_j, ring[j]);
+        }
+        // Add this ring to e0
+        extend_data(e0_data, R_i_j);
+    }
+    extend_data(e0_data, M);
+    // Hash data to produce e0 value
+    out.e = sha256_hash(e0_data);
+
+    // ---------------------------------------------------------------------
+    // Step 2: join up each ring at the index where we know the secret key
+    // ---------------------------------------------------------------------
+    for (size_t i = 0; i < rings.size(); ++i)
+    {
+        BITCOIN_ASSERT(i < known_key_indexes.size());
+        BITCOIN_ASSERT(i < k.size());
+        BITCOIN_ASSERT(i < out.s.size());
+
+        // Current ring and index of known key
+        const auto& ring = rings[i];
+        const size_t known_key_index = known_key_indexes[i];
+
+        // Calculate starting e value of this current ring.
+        auto e_i_j = borromean_hash(M, out.e, i, 0);
+
+        // Loop until index of known key.
+        for (size_t j = 0; j < known_key_index; ++j)
+        {
+            BITCOIN_ASSERT(j < out.s[i].size());
+            const auto& s = out.s[i][j];
+
+            // Calculate e and R until we reach our index.
+            BITCOIN_ASSERT(j < ring.size());
+            const auto R_i_j = calculate_R(s, e_i_j, ring[j]);
+            e_i_j = borromean_hash(M, R_i_j, i, j + 1);
+        }
+
+        // Find secret key used for calculation in the next step.
+        BITCOIN_ASSERT(known_key_index < ring.size());
+        const auto& known_public_key = ring[known_key_index];
+        BITCOIN_ASSERT(secret_keys.find(known_public_key) != secret_keys.end());
+        const auto& secret = secret_keys[known_public_key];
+
+        // Now close the ring using this calculation:
+        // s = k - e x
+        out.s[i][known_key_index] = calculate_s(k[i], e_i_j, secret);
+    }
+
+    return true;
+}
+
+bool verify(
+    const key_rings& rings,
+    const data_slice message,
+    const ring_signature& signature)
+{
+    // Compute message digest M
+    const data_chunk message_data = concatenate(message, rings);
+    const hash_digest M = sha256_hash(message_data);
+
+    // As compared with signing, we only have to perform a single step.
+    // The ring has already been computed, so now we just need to verify
+    // it by calculating e0, and looping all the way until the end R value
+    // which we use to recalculate e0.
+    // If the values match then we have a valid ring signature.
+
+    data_chunk e0_data;
+
+    // Loop through rings
+    for (size_t i = 0; i < rings.size(); ++i)
+    {
+        BITCOIN_ASSERT(i < signature.s.size());
+
+        const auto& ring = rings[i];
+
+        // Calculate first e value for this ring.
+        auto e_i_j = borromean_hash(M, signature.e, i, 0);
```

Comment:
From the paper (unknown if correct), e_{i, 0} = e_{0} which is a parameter of \theta equivalent to signature.e.  If that's correct, is there a reason to rehash instead of copy signature.e?

Reply:
I need to check the paper, but that's what the secp256k1-zkp code is doing too https://github.com/ElementsProject/secp256k1-zkp/blob/secp256k1-zkp/src/modules/rangeproof/borromean_impl.h#L81